### PR TITLE
(refactor)  - source code to ts

### PR DESCRIPTION
**Pull request by JoviDeCroock**

### CODE REVIEW

**Comment by jnwng** [comment]
qq — are our test suites running the typescript tests?

**Comment by JoviDeCroock** [comment]
@jnwng no, not yet. I first wanted to have approval on this part if you don't mind

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+// Note that this could do a lot more, such as reorder fields etc.
+function normalize(text: string): string {
+  return text.replace(/[\s,]+/g, ' ').trim();
+}
+
+type DocCache = { [docString: string]: DocumentNode };
+let docCache: DocCache = {};
+
+interface FragmentSourceMap {
+  [framentName: string]: {
```
Comment (jnwng) [inline src/index.ts]:
```suggestion
  [fragmentName: string]: {
```

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    [source: string]: boolean;
+  };
+}
+
+// A map fragmentName -> [normalized source]
+let fragmentSourceMap: FragmentSourceMap = {};
+
+function cacheKeyFromLoc(loc?: Location): string {
```
Comment (jnwng) [inline src/index.ts]: just for the sake of the translation, i dont think we should be changing the behavior of the individual functions just yet. any potential improvements _should_ be marked, but im afraid that this is going to have a cascading effect on existing behavior that we can't keep track of.

Comment (jnwng) [reply src/index.ts]: here's the source:
> function cacheKeyFromLoc(loc) {
  return normalize(loc.source.body.substring(loc.start, loc.end));
}

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+let printFragmentWarnings = true;
+function processFragments(ast: DocumentNode): DocumentNode {
+  const astFragmentMap: FragmentMap = {};
+  const definitions: DefinitionNode[] = [];
+
+  for (let i = 0; i < ast.definitions.length; i++) {
+    const fragmentDefinition = ast.definitions[i];
+
+    if (fragmentDefinition.kind === 'FragmentDefinition') {
+      const fragmentName: string = fragmentDefinition.name.value;
```
Comment (jnwng) [inline src/index.ts]: was this explicitly needed? does `fragmentDefinition.name.value` not expose the `string` type?

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      const sourceKey: string = cacheKeyFromLoc(fragmentDefinition.loc);
```
Comment (jnwng) [inline src/index.ts]: this equivalently should be inferred, correct?

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+export function disableFragmentWarnings(): void {
+  printFragmentWarnings = false;
+}
+
+function stripLoc(doc: Array<DocumentNode> | DocumentNode, removeLocAtThisLevel: number | boolean): DocumentNode  {
+  const docType = Object.prototype.toString.call(doc);
+
+  if (docType === '[object Array]' && doc instanceof Array) {
+    // This is wrong now because we return a DocumentNode... How do we approach this?
```
Comment (jnwng) [inline src/index.ts]:
```suggestion
    // TODO: This is wrong now because we return a DocumentNode... How do we approach this?
```

Comment (jnwng) [reply src/index.ts]: yeah, this looks a little strange, but in this case the return value of this function is not accurate — when we get an array of DocumentNode, we return an array of DocumentNode

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    return doc.map((d: DocumentNode) => stripLoc(d, removeLocAtThisLevel)) as any;
+  }
+
+  if (docType !== '[object Object]') {
+    throw new Error('Unexpected input.');
+  }
+
+  doc = doc as DocumentNode;
```
Comment (jnwng) [inline src/index.ts]: why do we have to recast this?

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+  // We don't want to remove the root loc field so we can use it
+  // for fragment substitution (see below)
+  if (removeLocAtThisLevel && doc.loc) {
+    // TODO: TS Does not allow deleting a read-only prop
+    delete (doc as any).loc;
```
Comment (jnwng) [inline src/index.ts]: sounds like we ought to make a mutable version of this...

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        // TODO: type this, need some more insight
+        (doc as any)[keys[key]] = stripLoc(value, true);
+      }
+    }
+  }
+
+  return doc;
+}
+
+let experimentalFragmentconstiables = false;
```
Comment (jnwng) [inline src/index.ts]:
```suggestion
let experimentalFragmentVariables = false;
```

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+function parseDocument(doc: string): DocumentNode {
+  const cacheKey = normalize(doc);
+
+  if (docCache[cacheKey]) {
+    return docCache[cacheKey];
+  }
+
+  // TODO: TS asked to change this from "experimentalFragmentconstiables" to "experimentalFragmentVariables"
```
Comment (jnwng) [inline src/index.ts]: i think your conversion from `var` to `const` was too greedy. making the adjustment above should resolve this

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+  // check that all "new" fragments inside the documents are consistent with
+  // existing fragments of the same name
+  parsed = processFragments(parsed);
+  parsed = stripLoc(parsed, false);
+  docCache[cacheKey] = parsed;
+
+  return parsed;
+}
+
+export function enableExperimentalFragmentconstiables(): void {
```
Comment (jnwng) [inline src/index.ts]:
```suggestion
export function enableExperimentalFragmentVariables(): void {
```

File: src/index.ts [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+  experimentalFragmentconstiables = true;
```
Comment (jnwng) [inline src/index.ts]:
```suggestion
  experimentalFragmentVariables = true;
```

File: src/index.ts [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+export function disableExperimentalFragmentconstiables(): void {
```
Comment (jnwng) [inline src/index.ts]:
```suggestion
export function disableExperimentalFragmentVariables(): void {
```

File: src/index.ts [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+  experimentalFragmentconstiables = false;
```
Comment (jnwng) [inline src/index.ts]:
```suggestion
  experimentalFragmentVariables = false;
```

File: src/index.ts [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+// XXX This should eventually disallow arbitrary string interpolation, like Relay does
+function gql(...args: Array<any>): DocumentNode {
+  const literals: any = args[0];
```
Comment (jnwng) [inline src/index.ts]: this should be inferred, correct?

File: src/index.ts (hunk 4 above)
Comment (JoviDeCroock) [reply src/index.ts]: Yes, will correct this

File: src/index.ts (hunk 6 above)
Comment (JoviDeCroock) [reply src/index.ts]: because otherwise it is an array or a node and here we are sure it's a node.

File: src/index.ts (hunk 7 above)
Comment (JoviDeCroock) [reply src/index.ts]: I have no experience with this, will look into it

File: src/index.ts (hunk 14 above)
Comment (JoviDeCroock) [reply src/index.ts]: correct

**Comment by JoviDeCroock** [comment]
Thanks for the extensive review, I'll make work of this tomorrow and also try to start the tests

File: src/index.ts (hunk 2 above)
Comment (JoviDeCroock) [reply src/index.ts]: It is the same as the source afaik?

**Comment by JoviDeCroock** [comment]
Problem being now that since we potentially return DocumentNode[] it kind off disrupts the tests

**Comment by abdonrd** [comment]
@jnwng any news here? Thanks in advance!

**Comment by JoviDeCroock** [comment]
@abdonrd I’ll likely close my apollo PR’s since I won’t have the time to correct PR remarks anymore. Feel free to fork my repo to get the code if you want to work on it.

**Comment by abdonrd** [comment]
@jnwng is there any plan to continue this?

**Pull request closed without merging**